packages/lib-runtime/src/store/coderefs.ts

```diff
@@ -0,0 +1,150 @@
+  typeof obj === 'function' &&
+  _.isEqual(Object.getOwnPropertySymbols(obj), [codeRefSymbol]) &&
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  (obj as any)[codeRefSymbol] === true;
+
+/**
+ * Parse data from the {@link EncodedCodeRef} object.
+ *
+ * Returns `undefined` if the `$codeRef` value is malformed.
+ */
+const parseEncodedCodeRef = (
+  ref: EncodedCodeRef,
+): [moduleName: string, exportName: string] | undefined => {
+  const match = ref.$codeRef.match(/^([^.]+)(?:\.([^.]+)){0,1}$/);
+
+  if (!match) {
+    return undefined;
+  }
+
+  const moduleName = match[1];
+  const exportName = match[2] || 'default';
+
+  if (!moduleName || !exportName) {
+    return undefined;
+  }
+
+  if (moduleName !== moduleName.trim() || exportName !== exportName.trim()) {
+    return undefined;
+  }
+
+  return [moduleName, exportName];
+};
+
+/**
+ * Create a {@link CodeRef} function from the {@link EncodedCodeRef} object.
+ *
+ * The referenced object is resolved just once. Multiple {@link CodeRef} Promise resolutions
+ * yield the exact same value.
+ */
+const createCodeRef = (
+  encodedRef: EncodedCodeRef,
+  entryModule: PluginEntryModule,
+  formatErrorMessage: (message: string) => string = _.identity,
+): CodeRef => {
+  let referencedObject: unknown;
+  let referencedObjectResolved = false;
+
+  const codeRef: CodeRef = async () => {
+    if (referencedObjectResolved) {
+      return referencedObject;
+    }
+
+    const refData = parseEncodedCodeRef(encodedRef);
+
+    if (!refData) {
+      throw new CodeRefError(
+        formatErrorMessage(`Malformed code reference '${encodedRef.$codeRef}'`),
+      );
+    }
+
+    const [moduleName, exportName] = refData;
+    let referencedModule: AnyObject;
+
+    try {
+      const moduleFactory = await entryModule.get(moduleName);
+      referencedModule = moduleFactory();
+    } catch (e) {
+      throw new CodeRefError(formatErrorMessage(`Failed to load module '${moduleName}'`), e);
+    }
+
+    if (!_.has(referencedModule, exportName)) {
+      throw new CodeRefError(
+        formatErrorMessage(`Missing module export '${moduleName}.${exportName}'`),
+      );
+    }
+
+    referencedObject = referencedModule[exportName];
+    referencedObjectResolved = true;
+
+    return referencedObject;
+  };
+
+  // Mark the function with a non-configurable symbol property
+  Object.defineProperty(codeRef, codeRefSymbol, { value: true });
+
+  return codeRef;
+};
+
+/**
+ * In the extension's `properties` object, replace all {@link EncodedCodeRef} values
+ * with {@link CodeRef} functions that can be used to load the referenced objects.
+ */
+export const decodeExtensionCodeRefs = (
+  extension: LoadedExtension,
+  entryModule: PluginEntryModule,
+) => {
+  visitDeep<EncodedCodeRef>(extension.properties, isEncodedCodeRef, (encodedRef, key, obj) => {
+    const codeRef = createCodeRef(
+      encodedRef,
+      entryModule,
+      (message) => `${message} in extension ${extension.uid}`,
```

Comment:
This adds useful diagnostics to `CodeRef` value resolution errors, so that developers know exactly from where they are coming from.